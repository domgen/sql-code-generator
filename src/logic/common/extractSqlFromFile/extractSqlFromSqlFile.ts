import { readFileAsync } from '../readFileAsync';

export const extractSqlFromSqlFile = ({ filePath }: { filePath: string }) => readFileAsync({ filePath }); // the sql is everything in the file in this case
