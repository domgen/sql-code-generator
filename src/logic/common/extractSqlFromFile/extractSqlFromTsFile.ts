import { readFileAsync } from '../readFileAsync';

/*
  note: in order to not block users when they have typescript type errors (by using ts-node),
    we can simply parse the file contents and grab the `export const sql = `__SQL__`;` w/ regex
*/
export const extractSqlFromTsFile = async ({ filePath }: { filePath: string }) => {
  // grab file contents
  const content = await readFileAsync({ filePath });

  // grab the sql definition w/ regex
  const [
    _, // tslint:disable-line no-unused
    sqlMatch,
  ] = new RegExp(/(?:export const sql = `)((.|\n)+)(?:`)/).exec(content) ?? [];

  // throw error if could not extract
  if (!sqlMatch) throw new Error(`could not extract sql from file at path '${filePath}'`);

  // return the sql
  return sqlMatch;
};
