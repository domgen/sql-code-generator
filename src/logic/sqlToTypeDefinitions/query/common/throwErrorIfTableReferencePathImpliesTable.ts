/**
 * instead of "select id from user" we're asking the user to "select user.id from user" or "select u.id from user u"
 *
 * rationale:
 * - this makes it so that we don't have to try and figure out the implicit reference
 * - this makes it so that when a user does join to another table, they don't have to add the explicit references after the fact (i.e., makes it easier to extend their queries)
 */
export const throwErrorIfTableReferencePathImpliesTable = ({ referencePath }: { referencePath: string }) => {
  if (referencePath.split('.').length !== 2) {
    throw new Error(
      `table reference of "${referencePath}" does not have the table alias explicitly defined, violating best practice`,
    );
  }
};
