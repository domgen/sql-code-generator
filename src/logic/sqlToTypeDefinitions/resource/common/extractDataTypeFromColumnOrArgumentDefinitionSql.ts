import { DataType } from '../../../../model';

// https://dev.mysql.com/doc/refman/8.0/en/string-types.html
const mysqlStringTypes = [
  'CHAR',
  'VARCHAR',
  'TINYTEXT',
  'TEXT',
  'MEDIUMTEXT',
  'LONGTEXT',
  'ENUM', // note: we treat enums as strings; TODO: https://github.com/uladkasach/sql-code-generator/issues/1
];

// https://dev.mysql.com/doc/refman/8.0/en/numeric-types.html
const mysqlNumberTypes = [
  'INTEGER',
  'INT',
  'SMALLINT',
  'TINYINT',
  'MEDIUMINT',
  'BIGINT',
  'DECIMAL',
  'NUMERIC',
  'FLOAT',
  'DOUBLE',
];

// https://dev.mysql.com/doc/refman/8.0/en/date-and-time-types.html
const mysqlDateTypes = ['DATE', 'TIME', 'DATETIME', 'TIMESTAMP', 'YEAR'];

// https://dev.mysql.com/doc/refman/8.0/en/binary-varbinary.html
const mysqlBinaryTypes = ['BINARY', 'VARBINARY'];

export const extractDataTypeFromColumnOrArgumentDefinitionSql = ({ sql }: { sql: string }) => {
  // 1. cast to upper
  const sqlUpper = sql.toUpperCase();

  // 2. one by one try to see which datatype is being defined in this definition
  if (mysqlStringTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.STRING;
  if (mysqlNumberTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.NUMBER;
  if (mysqlDateTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.DATE;
  if (mysqlBinaryTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.BUFFER;
  throw new Error(`could not extract data type from '${sql}'`);
};
