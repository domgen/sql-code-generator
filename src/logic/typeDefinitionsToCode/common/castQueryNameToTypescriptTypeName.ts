import { pascalCase } from 'pascal-case';

export const castQueryNameToTypescriptTypeName = ({
  name,
}: {
  name: string;
}) => {
  const queryNameInPascalCase = pascalCase(name);
  return `SqlQuery${queryNameInPascalCase}`;
};
