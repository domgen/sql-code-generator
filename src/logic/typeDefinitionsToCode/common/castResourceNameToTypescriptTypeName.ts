import { pascalCase } from 'pascal-case';

import { ResourceType } from '../../../model';

export const castResourceNameToTypescriptTypeName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  const resourceTypeInTitleCase = resourceType[0].toUpperCase() + resourceType.substr(1).toLowerCase();
  const resourceNameInPascalCase = pascalCase(name);
  return `Sql${resourceTypeInTitleCase}${resourceNameInPascalCase}`;
};
