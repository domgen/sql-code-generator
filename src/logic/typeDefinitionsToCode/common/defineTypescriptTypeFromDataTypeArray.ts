import { DataType } from '../../../model';

export const defineTypescriptTypeFromDataTypeArray = ({ type }: { type: DataType[] }) => type.join(' | ');
