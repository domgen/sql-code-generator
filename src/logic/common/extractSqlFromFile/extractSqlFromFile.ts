import { extractSqlFromSqlFile } from './extractSqlFromSqlFile';
import { extractSqlFromTsFile } from './extractSqlFromTsFile';

const SUPPORTED_FILE_EXTENSIONS = ['.sql', '.ts'];
export const extractSqlFromFile = async ({
  filePath,
}: {
  filePath: string;
}) => {
  // 1. grab extension from file; check that we support it
  const extension = `.${filePath.split('.').slice(-1)[0]}`;
  if (!SUPPORTED_FILE_EXTENSIONS.includes(extension)) {
    throw new Error(`file extension '.${extension}' is not supported`);
  }

  // 2. grab sql from file, based on file type
  if (extension === '.sql') return extractSqlFromSqlFile({ filePath });
  if (extension === '.ts') return extractSqlFromTsFile({ filePath });
  throw new Error('unexpected'); // should never reach here, since we checked earlier that the extension was supported
};
