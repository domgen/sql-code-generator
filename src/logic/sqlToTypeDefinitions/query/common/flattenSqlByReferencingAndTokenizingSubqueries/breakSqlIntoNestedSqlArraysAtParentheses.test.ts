import { extractSqlFromFile } from '../../../../common/extractSqlFromFile';
import { breakSqlIntoNestedSqlArraysAtParentheses } from './breakSqlIntoNestedSqlArraysAtParentheses';

describe('breakSqlIntoNestedSqlArraysAtParentheses', () => {
  it('should accurately break up this example', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../__test_assets__/find_with_subselect_in_select_expressions.sql`,
    });
    const nestedSqlArray = breakSqlIntoNestedSqlArraysAtParentheses({ sql });
    expect(nestedSqlArray).toHaveLength(3);
    expect(nestedSqlArray).toMatchSnapshot();
  });
  it('should retain the parentheses inside of the nested sql arrays', () => {
    const sql = "SELECT CONCAT('hel', 'lo');";
    const nestedSqlArray = breakSqlIntoNestedSqlArraysAtParentheses({ sql });
    expect(nestedSqlArray).toHaveLength(3);
    expect(nestedSqlArray[1][0][0]).toEqual('(');
    expect(nestedSqlArray[1][0].slice(-1)[0]).toEqual(')');
  });
});
