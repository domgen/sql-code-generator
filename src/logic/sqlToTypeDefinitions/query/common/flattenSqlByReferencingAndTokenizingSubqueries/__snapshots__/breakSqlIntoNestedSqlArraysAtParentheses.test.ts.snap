// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`breakSqlIntoNestedSqlArraysAtParentheses should accurately break up this example 1`] = `
Array [
  "-- query_name = find_with_subselect_in_select_expressions
SELECT
  s.id,
  s.uuid,
  s.name,
  ",
  Array [
    "(
    SELECT GROUP_CONCAT",
    Array [
      "(ice_cream_to_ingredient.ingredient_id ORDER BY ice_cream_to_ingredient.ingredient_id ASC SEPARATOR ',')",
    ],
    "
    FROM ice_cream_to_ingredient WHERE ice_cream_to_ingredient.ice_cream_id = s.id
  )",
  ],
  " as ingredient_ids,
  s.created_at
FROM ice_cream s
;
",
]
`;
